internal/provider/kubernetes/controller.go

```diff
@@ -0,0 +1,488 @@
+	}
+
+	// Watch Gateway CRUDs and reconcile affected GatewayClass.
+	if err := c.Watch(
+		&source.Kind{Type: &gwapiv1b1.Gateway{}},
+		handler.EnqueueRequestsFromMapFunc(r.processGateway),
+	); err != nil {
+		return err
+	}
+	if err := addGatewayIndexers(ctx, mgr); err != nil {
+		return err
+	}
+
+	// Watch HTTPRoute CRUDs and process affected Gateways.
+	if err := c.Watch(
+		&source.Kind{Type: &gwapiv1b1.HTTPRoute{}},
+		handler.EnqueueRequestsFromMapFunc(r.processHTTPRoute),
+	); err != nil {
+		return err
+	}
+	if err := addHTTPRouteIndexers(ctx, mgr); err != nil {
+		return err
+	}
+
+	// Watch TLSRoute CRUDs and process affected Gateways.
+	if err := c.Watch(
+		&source.Kind{Type: &gwapiv1a2.TLSRoute{}},
+		handler.EnqueueRequestsFromMapFunc(r.processTLSRoute),
+	); err != nil {
+		return err
+	}
+	if err := addTLSRouteIndexers(ctx, mgr); err != nil {
+		return err
+	}
+
+	// Watch Service CRUDs and process affected *Route objects.
+	if err := c.Watch(
+		&source.Kind{Type: &corev1.Service{}},
+		handler.EnqueueRequestsFromMapFunc(r.processService),
+	); err != nil {
+		return err
+	}
+
+	// Watch Secret CRUDs and process affected Gateways.
+	if err := c.Watch(
+		&source.Kind{Type: &corev1.Secret{}},
+		handler.EnqueueRequestsFromMapFunc(r.processSecret),
+	); err != nil {
+		return err
+	}
+
+	// Watch ReferenceGrant CRUDs and process affected Gateways.
+	if err := c.Watch(
+		&source.Kind{Type: &gwapiv1a2.ReferenceGrant{}},
+		handler.EnqueueRequestsFromMapFunc(r.processReferenceGrant),
+	); err != nil {
+		return err
+	}
+
+	r.log.Info("watching gatewayAPI related objects")
+	return nil
+}
+
+func (r *gatewayAPIReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.log.WithName(request.Name).Info("reconciling gatewayclass")
+
+	var gatewayClasses gwapiv1b1.GatewayClassList
+	if err := r.client.List(ctx, &gatewayClasses); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error listing gatewayclasses: %v", err)
+	}
+
+	var cc controlledClasses
+
+	for i := range gatewayClasses.Items {
+		if gatewayClasses.Items[i].Spec.ControllerName == r.classController {
+			// The gatewayclass was marked for deletion and the finalizer removed,
+			// so clean-up dependents.
+			if !gatewayClasses.Items[i].DeletionTimestamp.IsZero() &&
+				!slice.ContainsString(gatewayClasses.Items[i].Finalizers, gatewayClassFinalizer) {
+				r.log.Info("gatewayclass marked for deletion")
+				cc.removeMatch(&gatewayClasses.Items[i])
+				// Delete the gatewayclass from the watchable map.
+				r.resources.GatewayClasses.Delete(request.Name)
+				continue
+			}
+
+			cc.addMatch(&gatewayClasses.Items[i])
+		}
+	}
+
+	// The gatewayclass was already deleted/finalized and there are stale queue entries.
+	acceptedGC := cc.acceptedClass()
+	if acceptedGC == nil {
+		r.log.Info("failed to find an accepted gatewayclass")
+		return reconcile.Result{}, nil
+	}
+
+	// Store the accepted gatewayclass in the watchable map.
```

Comment:
this logic will need to change.
you will need to refactor https://github.com/envoyproxy/gateway/blob/54d44d57df53bd9ea9dec77f08a51fdbb0b28ed9/internal/message/types.go#L20 so it looks more like https://github.com/envoyproxy/gateway/blob/54d44d57df53bd9ea9dec77f08a51fdbb0b28ed9/internal/gatewayapi/translator.go#L51
```
import github.com/envoyproxy/gateway/internal/gatewayapi
```
type ProviderResources struct {
	watchable.Map[string, *gatewayapi.Resources]
}
```

this will allow you to populate the a single struct `gatewayapi.Resources` as you parse the reconciled resources and the last step should be to perform a `Store` into the watchable map

Reply:
with this logic you create a new struct every time the controller reconciles, and use the same key, moving the `diff`
complexity into the watchable layer which will notify the next layer (gateway api) if anything has changed or not

Reply:
it makes sense yes. I will load the obj into the resource table only at the gatewayclass reconcile level, and not at individual watch handlers.

Reply:
you would need to perform a `Store` at the end of reconciling all resources, so rather than having individual controllers, I recommend having one controller 
```
func gatewayAPIController() error {
   var res resources
   // Populate matched gateway class
   res.GatewayClass = gwClass
   // Populate matching gateways
   res.Gateways = append(res.Gateways, gw)
   // Populate Secrets
   .....
   pResources.Store("provider", res)
}
```

we need to perform the `Store` at the end to not cause any temporary churn in the data plane during restarts.